src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -518,71 +530,58 @@ public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompl
                 throw;
             }
 
-            return completionOption == HttpCompletionOption.ResponseContentRead && !string.Equals(request.Method.Method, "HEAD", StringComparison.OrdinalIgnoreCase) ?
-                FinishSendAsyncBuffered(sendTask, request, cts, disposeCts, cancellationToken, timeoutTime) :
-                FinishSendAsyncUnbuffered(sendTask, request, cts, disposeCts, cancellationToken, timeoutTime);
+            bool buffered = completionOption == HttpCompletionOption.ResponseContentRead &&
+                            !string.Equals(request.Method.Method, "HEAD", StringComparison.OrdinalIgnoreCase);
+
+            return FinishSendAsync(responseTask, request, cts, disposeCts, buffered, async, cancellationToken, timeoutTime);
         }
 
-        private async Task<HttpResponseMessage> FinishSendAsyncBuffered(
-            Task<HttpResponseMessage> sendTask, HttpRequestMessage request, CancellationTokenSource cts, bool disposeCts, CancellationToken callerToken, long timeoutTime)
+        private async ValueTask<HttpResponseMessage> FinishSendAsync(ValueTask<HttpResponseMessage> sendTask, HttpRequestMessage request, CancellationTokenSource cts,
+            bool disposeCts, bool buffered, bool async, CancellationToken callerToken, long timeoutTime)
         {
             HttpResponseMessage? response = null;
             try
             {
-                // Wait for the send request to complete, getting back the response.
-                response = await sendTask.ConfigureAwait(false);
-                if (response == null)
+                if (async)
                 {
-                    throw new InvalidOperationException(SR.net_http_handler_noresponse);
+                    // Wait for the send request to complete, getting back the response.
+                    response = await sendTask.ConfigureAwait(false);
                 }
-
-                // Buffer the response content if we've been asked to and we have a Content to buffer.
-                if (response.Content != null)
+                else
                 {
-                    await response.Content.LoadIntoBufferAsync(_maxResponseContentBufferSize, cts.Token).ConfigureAwait(false);
+                    // In sync scenario the ValueTask already contains the result, it has been constructed that way in Send method.
+                    response = sendTask.Result;
```

Comment:
It'd be better/simpler to just always await:
```C#
response = await sendTask.ConfigureAwait(false);
```
regardless of the value of `async`.  We want to make sync possible while minimizing any impact on the preferred async.

Reply:
Fixed, only added `Debug.Assert` for sync to check the task completeness.